splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/CrossJoinStrategy.java

```diff
@@ -0,0 +1,286 @@
+        // originalRestrictionList.setPredicatesAndProperties(storeRestrictionList);
+    }
+
+    @Override
+    public boolean doesMaterialization() {
+        return false;
+    }
+
+    /** @see JoinStrategy#multiplyBaseCostByOuterRows */
+	public boolean multiplyBaseCostByOuterRows() {
+		return true;
+	}
+
+    @Override
+    public OptimizablePredicateList getBasePredicates(OptimizablePredicateList predList, OptimizablePredicateList basePredicates, Optimizable innerTable) throws StandardException {
+        if (SanityManager.DEBUG) {
+            SanityManager.ASSERT(basePredicates.size() == 0,"The base predicate list should be empty.");
+        }
+
+        if (predList != null) {
+            predList.transferAllPredicates(basePredicates);
+            basePredicates.classify(innerTable, innerTable.getCurrentAccessPath().getConglomerateDescriptor());
+        }
+
+        /*
+         * We want all the join predicates to be included, so we just pass everything through and filter
+         * it out through the actual costing algorithm
+         */
+        return basePredicates;
+    }
+
+    @Override
+    public double nonBasePredicateSelectivity(Optimizable innerTable, OptimizablePredicateList predList) throws StandardException {
+        return 1.0;
+    }
+
+    @Override
+    public void putBasePredicates(OptimizablePredicateList predList, OptimizablePredicateList basePredicates) throws StandardException {
+        for (int i = basePredicates.size() - 1; i >= 0; i--) {
+            OptimizablePredicate pred = basePredicates.getOptPredicate(i);
+            predList.addOptPredicate(pred);
+            basePredicates.removeOptPredicate(i);
+        }
+    }
+
+	@Override
+	public boolean feasible(Optimizable innerTable,
+                            OptimizablePredicateList predList,
+                            Optimizer optimizer,
+                            CostEstimate outerCost,
+                            boolean wasHinted,
+                            boolean skipKeyCheck) throws StandardException {
+        AccessPath currentAccessPath = innerTable.getCurrentAccessPath();
+        boolean isSpark = false;
+        boolean isForcedControl = false;
+        boolean isOneRow = false;
+        boolean isHinted = currentAccessPath.isHintedJoinStrategy();
+        if (innerTable instanceof FromBaseTable) {
+            CompilerContext.DataSetProcessorType dspt = ((FromBaseTable) innerTable).getdataSetProcessorTypeForAccessPath(currentAccessPath);
+            isSpark = ((FromBaseTable)innerTable).isSpark(dspt);
+            isForcedControl = dspt.name().equals("FORCED_CONTROL");
+            isOneRow = ((FromBaseTable) innerTable).isOneRowResultSet();
+        }
+        // Only use cross join when it is inner join
+        // Only use cross join when it is on Spark, forced control and isHinted is for debug purpose
+        return !outerCost.isOuterJoin() && (isSpark || (isForcedControl && isHinted)) && !isOneRow;
```

Comment:
I see the difference. It makes it possible for the following query to go through broadcast join path:
```
splice> explain select * from --splice-properties joinOrder=fixed
>  t1, t2 --splice-properties joinStrategy=cross,useSpark=false
> ;
Plan                                                                                                                                                                                                                                                            
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Cursor(n=5,rows=400,updateMode=READ_ONLY (1),engine=control)                                                                                                                                                                                                    
  ->  ScrollInsensitive(n=4,totalCost=264.68,outputRows=400,outputHeapSize=1.23 KB,partitions=1)                                                                                                                                                                
    ->  CrossJoin(n=3,totalCost=96.68,outputRows=400,outputHeapSize=1.23 KB,partitions=1)                                                                                                                                                                       
      ->  TableScan[T2(1600)](n=2,totalCost=4.04,scannedRows=20,outputRows=20,outputHeapSize=1.23 KB,partitions=1)                                                                                                                                              
      ->  TableScan[T1(1584)](n=1,totalCost=4.04,scannedRows=20,outputRows=20,outputHeapSize=60 B,partitions=1)                                                                                                                                                 

5 rows selected
```
this query will go through broadcast join, while if we directly hint broadcast join, this error is reported:
```
splice> select * from --splice-properties joinOrder=fixed
>  t1, t2 --splice-properties joinStrategy=broadcast,useSpark=false
> ;
ERROR 42Y69: No valid execution plan was found for this statement. This is usually because an infeasible join strategy was chosen, or because an index was chosen which prevents the chosen join strategy from being used.
```
Is this the intention?

Reply:
In the first case, the query will still go to cross join, but the implementation is similar to broadcast join. In the second case, broadcast need to have hash key, so it cannot be choose.